### DISCUSSION TITLE
SERIAL/SSERIAL or TCPBridge (as serial) use to communicate with console (like WEB UI console)

### DISCUSSION THREAD

**Discussion by panciunio** [arendst/Tasmota#22195] [Ideas]
Hello,

Maybe it's not such an important feature, but I want to use Tasmota OS as an interface to communicate with MQTT using TTL serial communication. In short, I want to be able to type commands and receive responses on serial pins just like I can do in the Web UI console. I tried to do it using TCPbridge but after connection with Tasmota TCPbridge socket (via telnet) I can send only one line wchich can be interpreted and executed by Tasmota OS. No response and no possibility to send next string to OS.
Communication wiring:

RX - TX_TCPBridge (8888)
TX - RX_TCPBridge (8888)

Just two wires between particular pair.

Is possible to have simply communication with Tasmota OS using serial lines - doesn't matter hardware serial, software serial, TCPbridge... ?

Best Regards,
pancio

**Comment by sfromis** [comment]
Not clear what physical setup you are trying to make work, like what those "two wires" is connecting, and what you configure and do at each end.

In a small test setup, I had one Tasmota device with normal config (could be any serial device), another with TCP Tx/Rx (including doing a TCPStart), and connected the serial of both. Then I had no trouble with telnet into the 2nd Tasmota, issue multiple console commands/responses processed by the first.

**Reply by panciunio** [reply]
Thanks for reply,

Let's try to explain. I've dummy device like old terminal which has only TTL serial and I wantt to connect to Tasmota device in order to communicate with MQTT broker placed in my network. So from the dummy device perspective Tasmota device is a proxy:

![image](https://github.com/user-attachments/assets/f6a722d1-45cb-4425-9714-30de255de0a8)

My goal is to use Tasmota for communicate very old computer system with MQTT/Home Assistant and have to possibility to switch on/off lights, send message to direct device and so on. I thought that WEB UI CONSOLE is reflected on serial (any) but it doesn't... 

Best Regards,
pancio

**Reply by sfromis** [reply]
Warning: Many old devices with serial are using the standard RS232 with positive/negative 12V signals, not compatible with TTL 0/+3.3V supported by ESP8266. Hardware damage is a major risk.

If the hardware is suitable, and you configure serial pins for SerBr Rx/Tx, you can send data with the command `SerialSend` (use `SerialSend1` for automatic newline appended), and received data will be published as MQTT payloads. This is unrelated to the web UI console, but you can of course issue commands and watch responses there, and do the same via MQTT.

**Comment by panciunio** [comment]
Yes. I forgot to mention that I use a voltage shifter from 5V to 3V3. That's why I didn't use the word "RS232" :-)

1. I compiled own Tasmota 14.1.0 with:  `#define USE_TCP_BRIDGE` in in _user_config_override.h_ file, then flashed ESP32-WROOM-32D module (NodeMCU32). 
2. set pins in module:
![image](https://github.com/user-attachments/assets/59420183-9a6c-4c2e-9de2-13a742325121)
3. Added RULE1: `Rule1 ON System#Boot do TCPStart 8888 endon`
4. Acivated RULE1: `rule1 1`
5. After Restart I set TCPBaudRate: `TCPBaudRate 115200`
6. Restarted device...

After while I may connect to Serial bridge and saw connected client:

![image](https://github.com/user-attachments/assets/cc4cb56c-ab86-49e2-867f-fea030d8a819)

Everything looks good but...
 when I type 'rule1' on telnet window i see echo on SERIAL terminal and vice versa. Char by char.. so bridge working fine but still no reaction from Tasmota regarding message content...

![image](https://github.com/user-attachments/assets/cbe33ce9-7313-4267-bc30-4fe18b9b39db)

What I'm doing wrong?


EDIT1.
I see my mistake... shall be SerBrTX/RX.. not TCPTx/Rx...

EDIT2.
If I'm using RX0/TX0 device is unstable...
I checked configuration for RXD1/TXD1 and have no results... 


EDIT3. 
Finally I've positive result with transmission on RX2/TX2:

![image](https://github.com/user-attachments/assets/32af2bb6-3be5-4860-8691-5b455067586b)

but I still have no idea how to send data which can be interpreted by Tasmota.. e.g. `POWER1 ON`  

EDIT4.
Found relevant information:
https://tasmota.github.io/docs/Rules/#processing-json-received-from-softwareserialbridge

Will try...
 
Best Regards,
pancio
